### DISCUSSION TITLE
TPOT Logstash logs not forwarding to Splunk

### DISCUSSION THREAD

**Discussion by SAIKATDASGITHUB** [telekom-security/tpotce#1306] [Q&A]
- What version of the OS are you currently using `lsb_release -a` and `uname -a`? Debian bullseye
- What T-Pot version are you currently using? 22.04
- What edition (Standard, Nextgen, etc.) of T-Pot are you running?  Hive, Hive Sensor. Many Hive Sensors installed which is forwarding log to the central server Hive edition
- What architecture are you running on (i.e. hardware, cloud, VM, etc.)? Cloud
- Did you have any problems during the install? If yes, please attach `/install.log` `/install.err`. No
- How long has your installation been running? Many yrs
- Did you install upgrades, packages or use the update script? No
- Did you modify any scripts or configs? If yes, please attach the changes. No
- Please provide a screenshot of `glances` and `htop`. It is working
- How much free disk space is available (`df -h`)? Enough storage available
- What is the current container status (`dps.sh`)? All containers running
- What is the status of the T-Pot service (`systemctl status tpot`)? Active
- What ports are being occupied? Stop T-Pot `systemctl stop tpot` and run `netstat -tulpen` Basic ports used by Hive
- If a single container shows as `DOWN` you can run `docker logs <container-name>` for the latest log entries No containers down

We are running TPOT in a distributed environment. Many Hive Sensors editions are installed in different regions and in one central server Hive edition of TPOT is installed. All are running properly. 
We want to forward the logs from the central Hive edition to Splunk using HTTP Event Collector.

Steps we followed:
- Created HTTP token as mentioned in Splunk guide
-  Followed the steps mentioned here https://github.com/telekom-security/tpotce/wiki/Reconfigure-logstash.conf
-  Edited the logstash.conf file as mentioned here https://patroware.medium.com/how-to-forward-events-from-logstash-to-splunk-4f2608041feb
- Restarted TPOT
- Manually sent data through CURL is successful and the data is visible in Splunk
-  Logs are not forwarding to Splunk apart from the log sent thorough manual CURL
-  Logs in Kibana dashboard is properly visible

Please confirm if you need any specific logs and the path

/data/elk/logstash.conf

<img width="631" alt="Logstash Output" src="https://user-images.githubusercontent.com/116269856/230616675-b1b72c07-3a32-467d-ac82-701fac42f785.PNG">

/opt/tpot/etc/tpot.yml

<img width="434" alt="TPOT Logstash" src="https://user-images.githubusercontent.com/116269856/230613836-f0217d74-6a2a-4ec7-b8e2-2db352dff20c.PNG">

**Comment by t3chn0m4g3** [comment]
For Sensor installations the logic works a bit differently. Since we do not support other endpoints please check the configs and scripts for logstash [here](https://github.com/telekom-security/tpotce/blob/master/docker/elk/logstash/dist/entrypoint.sh) and make your adjustments accordingly.

**Comment by SAIKATDASGITHUB** [comment]
Can you plz tell what changes do we need to make in entrypoint.sh file to send the data to Splunk?

As the HTTP Event Collector was not working we installed Splunk Forwarder in the TPOT Server and monitored /data folder and all the logs from this folder is forwarding to Splunk. The data is indexing properly.

It will be beneficial to know HTTP Event Collector method of log forwarding as another method.

**Reply by ThomasToto** [reply]
I found a solution, I recommand you to modify the file "http_output.conf" located in /etc/elk/http_output.conf in logstash docker. 
So you have to follow "Reconfigure logstash.conf" but considering http_output.conf file instead of logstash.conf file. 
I specify that I am using Hive sensor and Hive installation. This modification has to be implemented on all Hive sensor. I didn't success to make it works with the Hive only.

**Reply by ishaughnessy3** [reply]
Are you replacing "%{log}" with anything? I've been playing around with different options but always end up with a string sent to splunk as opposed to a log entry.

**Reply by ThomasToto** [reply]
"%{log}" didn't work for me, I decided to do a little trick to put each log in a general json attribute. I created json object named _data_ and put all the data of each log in _donnee_ attribute.
Finally, I forward _donnee_ attribute to Splunk.
Here is the code : 
![Capture](https://github.com/telekom-security/tpotce/assets/58008070/f04d3695-c55a-49fa-8162-14070089aedc)

**Reply by SAIKATDASGITHUB** [reply]
@ThomasToto We will configure it in one of the sensor and test it.
